metricbeat/module/windows/perfmon/pdh_windows.go

```diff
@@ -0,0 +1,120 @@
+package perfmon
+
+import (
+	"strconv"
+	"unsafe"
+
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type Handle struct {
+	status      error
+	query       uintptr
+	counterType int
+	counters    []CounterGroup
+}
+
+type CounterGroup struct {
+	GroupName string
+	Counters  []Counter
+}
+
+type Counter struct {
+	counterName  string
+	counter      uintptr
+	counterPath  string
+	displayValue PdhCounterValue
+}
+
+var errorMapping = map[uint32]string{
+	PDH_INVALID_DATA:        `PDH_INVALID_DATA`,
+	PDH_INVALID_HANDLE:      `PDH_INVALID_HANDLE`,
+	PDH_NO_DATA:             `PDH_NO_DATA`,
+	PDH_NO_MORE_DATA:        `PDH_NO_MORE_DATA`,
+	PDH_STATUS_INVALID_DATA: `PDH_STATUS_INVALID_DATA`,
+	PDH_STATUS_NEW_DATA:     `PDH_STATUS_NEW_DATA`,
+	PDH_STATUS_NO_COUNTER:   `PDH_STATUS_NO_COUNTER`,
+	PDH_STATUS_NO_OBJECT:    `PDH_STATUS_NO_OBJECT`,
+}
+
+func GetHandle(config []CounterConfig) (*Handle, uint32) {
+	q := &Handle{}
+	err := _PdhOpenQuery(0, 0, &q.query)
+	if err != ERROR_SUCCESS {
+		return nil, err
+	}
+
+	counterGroups := make([]CounterGroup, len(config))
+	q.counters = counterGroups
+
+	for i, v := range config {
+		counterGroups[i] = CounterGroup{GroupName: v.Name, Counters: make([]Counter, len(v.Group))}
+		for j, v1 := range v.Group {
+			counterGroups[i].Counters[j] = Counter{counterName: v1.Alias, counterPath: v1.Query}
+			err := _PdhAddCounter(q.query, counterGroups[i].Counters[j].counterPath, 0, &counterGroups[i].Counters[j].counter)
+			if err != ERROR_SUCCESS {
+				return nil, err
+			}
+		}
+	}
+
+	return q, 0
+}
+
+func (q *Handle) ReadData(firstFetch bool) (common.MapStr, uint32) {
+
+	err := _PdhCollectQueryData(q.query)
+
+	if firstFetch {
+		// Most counters require two sample values in order to compute a displayable value. So wait and then collect the second value
+		time.Sleep(2000)
+		err = _PdhCollectQueryData(q.query)
+	}
+
+	if err != ERROR_SUCCESS {
+		return nil, err
+	}
+
+	result := common.MapStr{}
+
+	for _, v := range q.counters {
+		groupVal := make(map[string]interface{})
+		for _, v1 := range v.Counters {
+			err := _PdhGetFormattedCounterValue(v1.counter, PdhFmtDouble, q.counterType, &v1.displayValue)
+			if err != ERROR_SUCCESS {
+				return nil, err
+			}
+			doubleValue := (*float64)(unsafe.Pointer(&v1.displayValue.LongValue))
+			groupVal[v1.counterName] = *doubleValue
+
+		}
+		result[v.GroupName] = groupVal
+	}
+	return result, 0
+}
+
+func CloseQuery(q uintptr) uint32 {
+	err := _PdhCloseQuery(q)
+	if err != ERROR_SUCCESS {
+		return err
+	}
+
+	return 0
+}
+
+func GetError(err uint32) string {
```

Comment:
You are close to defining your own error type here. Consider doing something like this:

`type PdhError uint32`

Change `GetError` to `func (e PdhError) Error() string`. That makes `PdhError` satisfy the `error` interface and you and then return it from functions that user `error`.

Any place you have a uint32 error code you can cast it like `PdhError(2)`.